### PR TITLE
Priority Queue

### PR DESCRIPTION
**Pull request by ServiceDog**
See the updated SDD for information.

Overview:
- Redis queues used to store "tasks" to do
    - Different queues for emergency, high, medium, low priority tasks
    - Emergency tasks handled first and so on
    - Current tasks are "create_itinerary" and "cancel_itinerary"
    - Single threaded execution, currently no concurrent tasks. Option to expand here
- create_itinerary and cancel_itinerary now immediately return with a task_id, effectively a bakery ticket number
    - task status can be checked with get_task_status
    - queued tasks can be preemptively deleted with cancel_task
    - this means that clients don't block until the itinerary creation/deleted is completed (as in R3), asynchronous
- Replace confirm_itinerary with create_itinerary
    - No longer saving "draft" itineraries in local memory after a query
    - The client will submit the full itinerary to the scheduler (obtained from query_flight) instead of a draft ID number
    - "Reserving" multiple itineraries on a query was problematic, reduced options for other queries
    - Submitted itineraries undergo similar checks to query_flight but limited to the vertipads and aircraft listed in the itinerary, much faster
- Remove calls to svc-compliance
    - svc-compliance will periodically check storage for unapproved plans now

### CODE REVIEW

**Comment by ServiceDog** [comment]
image

**Review by owlot — COMMENTED** [review]
Great work introducing the Redis queues!
Just got a couple of comments 😁

File: server/src/tasks/cancel_itinerary.rs
[the remark this answers is not in the document]
Comment (ServiceDog) [reply server/src/tasks/cancel_itinerary.rs:1]: https://github.com/Arrow-air/svc-scheduler/commit/c23556655dbf758cb7c980b9e4dca5ec8798c8c6

File: server/src/tasks/create_itinerary.rs
[the remark this answers is not in the document]
Comment (ServiceDog) [reply server/src/tasks/create_itinerary.rs:1]: https://github.com/Arrow-air/svc-scheduler/commit/c23556655dbf758cb7c980b9e4dca5ec8798c8c6

File: server/src/tasks/cancel_itinerary.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    //
+    // TODO(R4): Heal the gap created by the removed flight plans
+    //
+
+    //
+    // Remove itinerary
+    //
+    let update_object = itinerary::UpdateObject {
+        id: itinerary_id.to_string(),
+        data: Option::from(itinerary::Data {
```
[the remark this answers is not in the document]
Comment (ServiceDog) [reply server/src/tasks/cancel_itinerary.rs]: https://github.com/Arrow-air/svc-scheduler/commit/c23556655dbf758cb7c980b9e4dca5ec8798c8c6

File: server/src/tasks/cancel_itinerary.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+pub async fn cancel_itinerary(task: &mut Task) -> Result<(), TaskError> {
+    let TaskBody::CancelItinerary(itinerary_id) = &task.body else {
+        tasks_error!("(cancel_itinerary) Invalid task details.");
+        return Err(TaskError::InvalidData);
+    };
+
+    tasks_info!("(cancel_itinerary) for id {}.", &itinerary_id);
+
+    let clients = get_clients().await;
+    let filter =
```
[the remark this answers is not in the document]
Comment (ServiceDog) [reply server/src/tasks/cancel_itinerary.rs]: https://github.com/Arrow-air/svc-scheduler/commit/c23556655dbf758cb7c980b9e4dca5ec8798c8c6

File: docs/icd.md [hunk 3]
```diff
@@ -46,6 +46,8 @@ See [High-Level Services ICD](https://github.com/Arrow-air/se-services/blob/deve
 | Service | Arguments | Description |
 | --- | --- | --- |
 | `is_ready` | (empty) | Returns `true` if server is up and running. |
-| `confirm_itinerary` | `string` id<br>`string` user_id | Takes `id` (UUID) of the draft itinerary and the user's UUID and confirms the flight. |
-| `cancel_itinerary`  | `string` id | Takes `id` (UUID) of the itinerary (either draft or confirmed) and cancels it. |
-| `query_flight` | `bool` isCargo<br/>`uint32` persons<br/>`uint32` weight_grams<br/>`Timestamp` earliest departure time<br/>`Timestamp` latest arrival time<br/>`string` vertiport_depart_id<br/>`string` vertiport_arrive_id<br/> | Takes requested departure and arrival vertiport UUIDs and a time window for the itinerary to occur and returns a number of draft itineraries. |
+| `create_itinerary` | TODO | Takes a proposal for an itinerary and creates it if possible. Returns a task ID immediately. |
+| `cancel_itinerary` | `string` id | Takes `id` (UUID) of an itinerary and cancels it. Returns a task ID immediately. |
+| `cancelTask` |  `uint32` task_id | Cancels a queued task before it can be acted on, if still queued. |
```
[the remark this answers is not in the document]
Comment (ServiceDog) [reply docs/icd.md]: https://github.com/Arrow-air/svc-scheduler/commit/c23556655dbf758cb7c980b9e4dca5ec8798c8c6

**Comment by ServiceDog** [comment]
image

Had to signoff CI commits due to a missing signature

**Comment by ServiceDog** [comment]
GitHub is struggling to understand that I modified docker-compose.yml when it was previously removed, though locally I have no merge conflicts. Just making a different PR: https://github.com/Arrow-air/svc-scheduler/pull/43

**Pull request closed without merging**